src/lib/snarky/src/snark0.ml

```diff
@@ -915,159 +875,134 @@ module Make_basic (Backend : Backend_intf.S) = struct
+                   if Option.is_some s && not (Constraint.eval c get_value)
+                   then
+                     failwithf "Constraint unsatisfied:\n%s\n%s\n"
+                       (Constraint.annotation c)
+                       (Constraint.stack_to_string stack)
+                       () ;
+                   Constraint.add ~stack c system )) ;
             go stack t handler s
         | With_state (p, and_then, t_sub, k) ->
-            let s, s_sub = As_prover.run p get_value s in
+            let s, s_sub = run_as_prover (Some p) s in
             let s_sub, y = go stack t_sub handler s_sub in
-            let s, () = As_prover.run (and_then s_sub) get_value s in
+            let s, _ = run_as_prover (Option.map ~f:and_then s_sub) s in
             go stack (k y) handler s
         | With_handler (h, t, k) ->
             let s', y = go stack t (Request.Handler.push handler h) s in
             go stack (k y) handler s'
         | Clear_handler (t, k) ->
             let s', y = go stack t Request.Handler.fail s in
             go stack (k y) handler s'
-        | Exists ({store; check; _}, p, k) ->
-            let s', value = Provider.run p get_value s handler in
-            let var = Typ.Store.run (store value) store_field_elt in
-            let (), () = go stack (check var) handler () in
-            go stack (k {Handle.var; value= Some value}) handler s'
+        | Exists ({store; alloc; check; _}, p, k) -> (
+          match s with
+          | Some s ->
+              let s', value = Provider.run p get_value s handler in
+              let var = Typ.Store.run (store value) store_field_elt in
+              (* TODO: Push a label onto the stack here *)
+              let _, () = go stack (check var) handler (Some ()) in
+              go stack (k {Handle.var; value= Some value}) handler (Some s')
+          | None ->
+              let var = Typ.Alloc.run alloc alloc_var in
+              (* TODO: Push a label onto the stack here *)
+              let _, () = go stack (check var) handler None in
+              go stack (k {Handle.var; value= None}) handler None )
         | Next_auxiliary k -> go stack (k !next_auxiliary) handler s
       in
-      match go [] t0 Request.Handler.fail s0 with
+      let retval = go [] t0 Request.Handler.fail s0 in
+      ignore
+        (Option.map system ~f:(fun system ->
+             let auxiliary_input_size = !next_auxiliary - (1 + num_inputs) in
+             R1CS_constraint_system.set_auxiliary_input_size system
+               auxiliary_input_size )) ;
+      retval
+
+    (* TODO-someday: Add pass to unify variables which have an Equal constraint *)
+    let constraint_system ~num_inputs (t : (unit, 's) t) :
+        R1CS_constraint_system.t =
+      let input = Field.Vector.create () in
+      let next_auxiliary = ref (1 + num_inputs) in
+      let aux = Field.Vector.create () in
```

Comment:
It would be nice if we could avoid having to construct this vector when we don't use it (just from a code-cleanliness perspective) but not a huge deal

Reply:
I tried to do this (2c4dce470fed68e33ef5f768116ffbd9d5f76f5e), but it makes a bit of a mess of `run`. I haven't included it here, but I can pull it in/try a different approach if necessary.